### PR TITLE
Unify kotlin compiler config for building Gradle and for the binary compatibility checks

### PR DESCRIPTION
**Pull request by eskatos**
There were some discrepancies.

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
@bot-gradle test this

File: build-logic/basics/src/main/kotlin/gradlebuild/basics/kotlindsl/kotlin-compiler-configuration.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.load.java.JavaTypeEnhancementState
+import org.jetbrains.kotlin.load.java.Jsr305Settings
+import org.jetbrains.kotlin.load.java.ReportLevel
+
+
+fun KotlinCompile.configureKotlinCompilerForGradleBuild() {
+    compilerOptions {
+        allWarningsAsErrors = true
+        apiVersion = KotlinVersion.KOTLIN_1_8
```
Comment (ljacomet) [inline build-logic/basics/src/main/kotlin/gradlebuild/basics/kotlindsl/kotlin-compiler-configuration.kt:40]: 🤔 Is there a way we could share the common values between the two extension methods?

Right now, this code remains problematic as someone could update a line in one function and not in the other.

Comment (eskatos) [reply build-logic/basics/src/main/kotlin/gradlebuild/basics/kotlindsl/kotlin-compiler-configuration.kt:40]: These are two different APIs and use different types.
It's already good that these two functions are next to each other in a file with nothing else.

Comment (ljacomet) [reply build-logic/basics/src/main/kotlin/gradlebuild/basics/kotlindsl/kotlin-compiler-configuration.kt:40]: Oh right, I focused on `KOTLIN_1_8 and missed they are no on the same types ...

**Review by ljacomet — APPROVED** [review]

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**